### DISCUSSION TITLE
Conditional Coalescing and Assignment Operators

### DISCUSSION THREAD

**Discussion by Ai-N3rd** [dotnet/csharplang#7897] [Language Ideas] [answered]
### Summary:
I would like to propose the addition of conditional assignment operators >?= and <?= to improve the expressiveness and conciseness of conditional assignments in C#.

### Motivation:
Currently, conditional assignments often involve an extra line of code or ternary operators, making the code less readable and more verbose. The proposed operators aim to enhance code readability and reduce verbosity in certain scenarios.

Syntax Proposal

```csharp
int x = 10;
int y = 5;

x >?= y; // Expands to x = x > y ? x : y
x <?= y; // Expands to x = x > y ? x : y

Console.WriteLine(x >? y); // Expands to Console.WriteLine(x > y ? x : y);
Console.WriteLine(x <? y); // Expands to Console.WriteLine(x < y ? x : y);
```

### Additional Information

**Impact on Existing Code:**
The proposed operators should not have any impact on existing code, as they are new additions.

**Alternatives Considered:**
The alternatives include using traditional conditional statements or ternary operators, but these may result in less readable and more verbose code.

**Conclusion**
The addition of conditional assignment operators would offer a concise and expressive way to handle conditional assignments, especially when dealing with minimum and maximum value assignments.

Edits: Fixed formatting mistakes

**Comment by HaloFour** [comment]
```cs
using static System.Math;

x = Max(x, y);
x = Min(x, y);
```

**Reply by jnm2** [reply]
Or extension methods:
```cs
x = x.Max(y);
x = x.Min(y);
```

**Reply by ufcpp** [reply]
```cs
using System.Numerics;

var x = 5;
var y = 3;

x.MaxBy(y);

Console.WriteLine(x);

static class Ex
{
    public static void MaxBy<T>(ref this T x, T y)
        where T: struct, INumber<T>
    {
        x = T.Max(x, y);
    }
}
```

**Reply by Ai-N3rd** [reply]
These do not account for operator overloading nor anything besides numbers.

**Reply by colejohnson66** [reply]
Are you saying @ufcpp's code won't work for non-number types? Why is that needed? What non-number type needs a "min" and "max" comparison? How would your proposal work for things that aren't numbers?

**Comment by theunrepentantgeek** [comment, marked as answer]
Making C# cryptic has never been a goal of the LDM folks. Including a `?` might be necessary to avoid collision with other operators, but there's nothing here about null tolerance, which is what `?` is often used for - so these would be very confusing symbols.

Reducing character count isn't _of itself_ a sufficient goal either.

Check out [-100 points](https://learn.microsoft.com/en-us/archive/blogs/ericgu/minus-100-points).
